src/main.ts

```diff
@@ -298,17 +300,50 @@ async function buildAndGetErrors(repoDir: string, tscPath: string, skipLibCheck:
     }
 }
 
-async function installPackages(repoDir: string, recursiveSearch: boolean, types?: string[]) {
-    const commands = await ip.restorePackages(repoDir, /*ignoreScripts*/ true, recursiveSearch, /*lernaPackages*/ undefined, types);
+async function installPackages(repoDir: string, recursiveSearch: boolean, timeoutMs: number, types?: string[]) {
     let usedYarn = false;
-    for (const { directory: packageRoot, tool, arguments: args } of commands) {
-        await new Promise<void>((resolve, reject) => {
-            usedYarn = usedYarn || tool === ip.InstallTool.Yarn;
-            cp.execFile(tool, args, { cwd: packageRoot }, err => err ? (((err as any).packageRoot = packageRoot), reject(err)) : resolve());
-        });
+    try {
+        const startMs = performance.now();
+        const commands = await ip.restorePackages(repoDir, /*ignoreScripts*/ true, recursiveSearch, /*lernaPackages*/ undefined, types);
+        for (const { directory: packageRoot, tool, arguments: args } of commands) {
+            let installError: any = undefined;
+
+            const elapsedMs = performance.now() - startMs;
+            // Fudge factor: don't start an installation with less than 100 ms left
+            // Reduces the odds of a race between starting the install process and killing it
+            if (elapsedMs > timeoutMs - 100) {
+                installError = new Error(`Timed out after ${timeoutMs} ms`);
+            }
+            else {
+                usedYarn = usedYarn || tool === ip.InstallTool.Yarn;
+
+                await new Promise<void>(resolve => {
+                    const timeout = setTimeout(async () => {
+                        installError ??= new Error(`Timed out after ${timeoutMs} ms`);
+
+                        await execAsync(processCwd, `./scripts/kill-children-of ${processPid} ${tool}`);
+                        resolve();
+                    }, timeoutMs - elapsedMs);
+
+                    cp.execFile(tool, args, { cwd: packageRoot }, err => {
+                        installError ??= err;
+                        clearTimeout(timeout);
+                        resolve();
+                    });
+                });
+            }
+
+            if (installError) {
+                // Attach the package root to improve error reporting
+                installError.packageRoot = packageRoot.substring(repoDir.length + 1) || "root directory";
```

Comment:
I believe you could attach packageRoot in both places you set `installError` and then call `reject(installError)` -- the code would be shorter and less surprising, at the cost of some duplication.

You could then eagerly calculate something like `errorPackageRoot` at the top of the loop to get rid of the duplication.

Reply:
I was over-indexing on not throwing and re-throwing.  Simplified.